<a href="https://colab.research.google.com/github/KOMBOU12/Marius/blob/main/Pr%C3%A9diction_conforme_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Projet : Classification et  Prédiction conforme**
Le jeu de données **CIFAR-10** contient 60 000 images en couleur de taille 32x32. Ces images représentent des objets appartenant à 10 classes :

*   avion
*   automobile
*   oiseau
*   chat
*   cerf
*   chien
*   grenouille
*   cheval
*   bateau
*   camion

Le jeu de données d'entraînement contient 50 000 images d'entraînement, soit 5 000 images par classe. Le jeu de test contient 10 000 images de test, soit 1 000 images par classe.

**Pourquoi ce jeu de donnée ?**

CIFAR-10 comme jeu de données est un bon point de départ pour entraîner et tester des Convolutional Neural Networks (CNN), car il n'est pas trop volumineux mais suffisamment varié pour évaluer la performance d'un modèle.

**Importation des données et statistiques simples**

Tu dois dire pourquoi tu choisis ce jeu de donnée. Tu dois également faire de la regression linéaire sur ce jeu de donnée et dire si ça marche ou pas et pourquoi. Vu que tu as une CNN tu peux peut être penser faire de la CV+,  pour faire de la prédiction conforme. Il semble que la SCP serait judicieux

In [2]:
from tensorflow.keras.datasets import cifar10
from sklearn.model_selection import train_test_split

# Charger les données CIFAR-10
(X_full, y_full), (X_test, y_test) = cifar10.load_data()

# Diviser les données en trois ensembles : entraînement, calibration et test
X_train, X_calibration, y_train, y_calibration = train_test_split(X_full, y_full, test_size=0.2, random_state=42)

# Normalisation des données
X_train = X_train.astype('float32') / 255.0
X_calibration = X_calibration.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0

170498071/170498071 ━━━━━━━━━━━━━━━━━━━━ 3s 0us/step


On entraînne le modèle

In [3]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

# Entraîner le modèle sur l'ensemble d'entraînement
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train, epochs=10, batch_size=64, validation_split=0.1)

# Prédire sur l'ensemble de calibration
y_pred_calibration = model.predict(X_calibration)



/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 50s 86ms/step - accuracy: 0.3312 - loss: 1.8235 - val_accuracy: 0.4980 - val_loss: 1.4141
Epoch 2/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 46s 82ms/step - accuracy: 0.5469 - loss: 1.2931 - val_accuracy: 0.5935 - val_loss: 1.1827
Epoch 3/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 92s 101ms/step - accuracy: 0.6182 - loss: 1.1028 - val_accuracy: 0.6148 - val_loss: 1.0842
Epoch 4/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 71s 82ms/step - accuracy: 0.6476 - loss: 1.0121 - val_accuracy: 0.6275 - val_loss: 1.0474
Epoch 5/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 50s 88ms/step - accuracy: 0.6796 - loss: 0.9274 - val_accuracy: 0.6375 - val_loss: 1.0359
Epoch 6/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 80s 85ms/step - accuracy: 0.6954 - loss: 0.8880 - val_accuracy: 0.6390 - val_loss: 1.0288
Epoch 7/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 83s 88ms/step - accuracy: 0.7086 - loss: 0.8349 - val_accuracy: 0.6718 - val_loss: 0.9620
Epoch 8/10
563/563 ━━━━━━━━━━━━━━━━━━━━ 79s 82ms/step - accuracy: 0.7296 - loss: 0.7774 -

# **La prédiction conforme**

On fait une SCP parceque nous avons beaucoup de données et on cherche à ajuster la distribution des prédictions et fournir un intervalle de confiance pour les prédictions sur les nouvelles données

1. Calcul des scores de non-conformité sur l'ensemble de calibration

In [4]:
import numpy as np

# Prédictions de calibration
y_pred_calibration_probs = model.predict(X_calibration)

# Extraire les probabilités de la vraie classe
true_class_probs = np.array([y_pred_calibration_probs[i, y_calibration[i]] for i in range(len(y_calibration))])

# Calculer les scores de non-conformité
non_conformity_scores = 1 - true_class_probs


313/313 ━━━━━━━━━━━━━━━━━━━━ 9s 29ms/step


2. On fixe le niveau de couverture souhaité

In [5]:
# Fixer le niveau de couverture à 90%
alpha = 0.1
quantile_threshold = np.quantile(non_conformity_scores, 1 - alpha)

3. On Applique le seuil de non-conformité sur les nouvelles prédictions

In [6]:
# Prédictions sur l'ensemble de test
y_pred_test_probs = model.predict(X_test)

# Construire l'ensemble prédictif pour chaque échantillon de test
prediction_sets = []
for i in range(len(X_test)):
    prediction_set = np.where(y_pred_test_probs[i] >= 1 - quantile_threshold)[0]
    prediction_sets.append(prediction_set)


313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step


4. On calcul la couverture et la taille moyenne des ensembles prédictifs

In [7]:
# Calculer la couverture
correct_cover = [y_test[i] in prediction_sets[i] for i in range(len(y_test))]
coverage = np.mean(correct_cover)

# Calculer la taille moyenne des ensembles prédictifs
average_set_size = np.mean([len(prediction_set) for prediction_set in prediction_sets])

print(f"Couverture : {coverage * 100:.2f}%")
print(f"Taille moyenne des ensembles prédictifs : {average_set_size:.2f}")


Couverture : 89.21%
Taille moyenne des ensembles prédictifs : 2.13




*   **Couverture de 89,21%**: Cela signifie que, dans environ 9 cas sur 10, la véritable classe est incluse dans l'ensemble prédictif proposé, ce qui est proche de l'objectif de 90%.
*   La **taille moyenne des ensembles prédictifs** indique le nombre moyen de classes incluses dans l'ensemble proposé pour chaque prédiction. Ici, chaque prédiction conforme propose en moyenne 2,13 classes, ce qui signifie que le modèle fournit un ensemble de 2 ou 3 classes en moyenne pour chaque image testée.



# **Regression Linéaire**

In [8]:
import numpy as np
from tensorflow.keras.datasets import cifar10
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

# Charger les données CIFAR-10
(X_full, y_full), (X_test, y_test) = cifar10.load_data()

# Redimensionner les données pour les rendre compatibles avec la régression linéaire
# Convertir les images de 32x32x3 en vecteurs de 32*32*3 = 3072 dimensions
X_full_flat = X_full.reshape((X_full.shape[0], -1))
X_test_flat = X_test.reshape((X_test.shape[0], -1))

# Diviser les données en ensembles d'entraînement et de calibration
X_train, X_calibration, y_train, y_calibration = train_test_split(X_full_flat, y_full, test_size=0.2, random_state=42)

# Normaliser les données pour que les pixels soient compris entre 0 et 1
X_train = X_train.astype('float32') / 255.0
X_calibration = X_calibration.astype('float32') / 255.0
X_test_flat = X_test_flat.astype('float32') / 255.0

# Appliquer une régression linéaire
model = LinearRegression()

# Entraîner le modèle sur l'ensemble d'entraînement
model.fit(X_train, y_train)

# Prédire sur l'ensemble de test
y_pred = model.predict(X_test_flat)

# Calculer l'erreur quadratique moyenne (MSE) et le R² pour évaluer les performances
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Erreur quadratique moyenne (MSE) : {mse}")
print(f"Coefficient de détermination (R²) : {r2}")


Erreur quadratique moyenne (MSE) : 8.177422523498535
Coefficient de détermination (R²) : 0.008797228336334229


Une MSE de 8.18 signifie que, en moyenne, les prédictions sont à une distance quadratique de 8.18 des vraies valeurs de classe. Puisque les classes vont de 0 à 9, cela montre que le modèle fait des erreurs assez importantes.

La **régression linéaire** n'est pas un modèle adapté pour ce type de données, car elle suppose une relation linéaire continue entre les pixels et les classes. Or, dans CIFAR-10, les classes sont des labels discrets (catégories d'objets comme des chats, avions, etc.), et il n'existe pas de relation continue ou ordinale entre elles.